dask/dataframe/rolling.py

```diff
@@ -91,31 +148,193 @@ def map_overlap(func, df, before, after, *args, **kwargs):
         ):
             raise ValueError("before and after must be positive integers")
 
-    if "token" in kwargs:
-        func_name = kwargs.pop("token")
-        token = tokenize(df, before, after, *args, **kwargs)
+    name = kwargs.pop("token", None)
+    parent_meta = kwargs.pop("parent_meta", None)
+
+    assert callable(func)
+    if name is not None:
+        token = tokenize(meta, *args, **kwargs)
     else:
-        func_name = "overlap-" + funcname(func)
-        token = tokenize(func, df, before, after, *args, **kwargs)
+        name = funcname(func)
+        token = tokenize(func, meta, *args, **kwargs)
+    name = f"{name}-{token}"
+
+    from dask.array.core import normalize_arg
+    from dask.dataframe.core import (
+        Scalar,
+        _get_divisions_map_partitions,
+        _get_meta_map_partitions,
+        _is_only_scalar,
+        _maybe_from_pandas,
+        apply_and_enforce,
+        new_dd_object,
+        partitionwise_graph,
+    )
+    from dask.dataframe.multi import _maybe_align_partitions
+    from dask.delayed import unpack_collections
+    from dask.utils import apply
```

Comment:
Shall we move all imports to the top level?

Reply:
Done